### PR TITLE
[New resource] TeamsProtectionPolicy

### LINKED ISSUE #5296: TeamsProtectionPolicy: New resource proposal
### Description
With TeamsProtectionPolicy the user can modify Microsoft Teams protection policies. This can be reached through Exchange Online.
This is the link for the Powershell version: https://learn.microsoft.com/en-us/powershell/module/exchange/set-teamsprotectionpolicy?view=exchange-ps

### Proposed properties
Set-TeamsProtectionPolicy
   [-Identity] <TeamsProtectionPolicyIdParameter>
   [-AdminDisplayName <String>] # The AdminDisplayName parameter specifies a description for the policy
   [-HighConfidencePhishQuarantineTag <String>] # The HighConfidencePhishQuarantineTag parameter specifies the quarantine policy that's used for 
   messages that are quarantined as high confidence phishing by ZAP for Teams.
   [-MalwareQuarantineTag <String>] # The MalwareQuarantineTag parameter specifies the quarantine policy that's used for messages that are 
   quarantined as malware by ZAP for Teams.
   [-ZapEnabled <Boolean>] # The ZapEnabled parameter specifies whether to enable zero-hour auto purge (ZAP) for malware and high confidence 
   phishing messages in Teams messages.

I think it would be a great addition for the Teams assets. 
Thank you! :)

### PR DESCRIPTION
**Pull request by ykuijs**
#### Pull Request (PR) description
This PR adds a new resource for TeamsProtectionPolicy

#### This Pull Request (PR) fixes the following issues
- Fixes #5296

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [x] Resource parameter descriptions added/updated in the schema.mof.
- [x] Resource documentation added/updated in README.md.
- [x] Resource settings.json file contains all required permissions.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated.
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

### CODE REVIEW

**Comment by ricmestre** [comment]
@ykuijs I understand why you called it TeamsProtectionPolicy but this is actually an EXO resource, I think it should be renamed to maybe EXOTeamsProtectionPolicy.

File: Modules/Microsoft365DSC/DSCResources/MSFT_TeamsProtectionPolicy/MSFT_TeamsProtectionPolicy.schema.mof [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    [Write, Description("The AdminDisplayName parameter specifies a description for the policy.")] String AdminDisplayName;
+    [Write, Description("The HighConfidencePhishQuarantineTag parameter specifies the quarantine policy that's used for messages that are quarantined as high confidence phishing by ZAP for Teams."), ValueMap{"AdminOnlyAccessPolicy","DefaultFullAccessPolicy","DefaultFullAccessWithNotificationPolicy"}, Values{"AdminOnlyAccessPolicy","DefaultFullAccessPolicy","DefaultFullAccessWithNotificationPolicy"}] String HighConfidencePhishQuarantineTag;
+    [Write, Description("The MalwareQuarantineTag parameter specifies the quarantine policy that's used for messages that are quarantined as malware by ZAP for Teams."), ValueMap{"AdminOnlyAccessPolicy","DefaultFullAccessPolicy","DefaultFullAccessWithNotificationPolicy"}, Values{"AdminOnlyAccessPolicy","DefaultFullAccessPolicy","DefaultFullAccessWithNotificationPolicy"}] String MalwareQuarantineTag;
+    [Write, Description("he ZapEnabled parameter specifies whether to enable zero-hour auto purge (ZAP) for malware and high confidence phishing messages in Teams messages.")] Boolean ZapEnabled;
```
Comment (ricmestre) [inline Modules/Microsoft365DSC/DSCResources/MSFT_TeamsProtectionPolicy/MSFT_TeamsProtectionPolicy.schema.mof]: Missing initial 'T'

File: Modules/Microsoft365DSC/Examples/Resources/TeamsProtectionPolicy/2-Update.ps1 [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    node localhost
+    {
+        TeamsProtectionPolicy "TeamsProtectionPolicy"
+        {
+            IsSingleInstance                 = "Yes"
+            AdminDisplayName                 = "Contoso Administrator"
+            HighConfidencePhishQuarantineTag = "DefaultFullAccessPolicy"
+            MalwareQuarantineTag             = "AdminOnlyAccessPolicy"
+            ZapEnabled                       = $true
+            Credential                       = $credsCredential
```
Comment (ricmestre) [inline Modules/Microsoft365DSC/Examples/Resources/TeamsProtectionPolicy/2-Update.ps1]: All examples must use certificate authentication, Nik requested me to change all of them a few months ago.

File: Modules/Microsoft365DSC/DSCResources/MSFT_TeamsProtectionPolicy/MSFT_TeamsProtectionPolicy.schema.mof (hunk 1 above)
Comment (ykuijs) [reply Modules/Microsoft365DSC/DSCResources/MSFT_TeamsProtectionPolicy/MSFT_TeamsProtectionPolicy.schema.mof]: Great spot....was a nice copy/paste error on my end :-)

File: Modules/Microsoft365DSC/Examples/Resources/TeamsProtectionPolicy/2-Update.ps1 (hunk 2 above)
Comment (ykuijs) [reply Modules/Microsoft365DSC/Examples/Resources/TeamsProtectionPolicy/2-Update.ps1]: Checked a few others and those were all also using creds. Corrected these as well, but that does mean we have to go through the entire list as well......or maybe create a unit test for it. For now this is solved!

**Comment by ykuijs** [comment]
> @ykuijs I understand why you called it TeamsProtectionPolicy but this is actually an EXO resource, I think it should be renamed to maybe EXOTeamsProtectionPolicy.

That was something I was thinking about as well when developing. We have to connect to ExchangeOnline and change the setting there, but the setting is for Teams.......I chose to name it Teams since it is really meant for Teams.

@NikCharlebois, what is your opinion on this?

**Comment by ricmestre** [comment]
I use separate app registrations for each workload and they have their separate API permissions/roles, it will be hard to justify my customers which have EXO and Teams workloads why they will need to have one EXO admin for EXO and another one for Teams just because of a single resource.

**Comment by ykuijs** [comment]
True, depends from which angle you look at it. From a management perspective it is confusing that if you want to manage Teams, you have to use Teams resources and a single EXO resource.

**Comment by ykuijs** [comment]
Just checked with Nik. There are more resources like this, where we are following the underlying API in the naming. Will update this resource as well to match the EXO naming.

**Pull request merged**